File: frontend/src/api/todo.ts
import { api } from "./axios";

export type TodoStatus = "completed" | "pending" | "incomplete";

export interface Todo {
  id: string;
  user_id: string;
  name: string;
  description: string;
  pending_at: string | null;
  completed_at: string | null;
  created_at: string;
  archived_at?: string | null;
}

export interface CreateTodoPayload {
  name: string;
  description?: string;
  pending_at?: string | null;
}

export interface UpdateTodoPayload {
  name?: string;
  description?: string;
  pending_at?: string | null;
  completed_at?: string | null;
}

export interface GetTodosParams {
  status?: TodoStatus;
  search?: string;
  page?: number;
  limit?: number;
}

export interface GetTodosResponse {
  page: number;
  limit: number;
  todos: Todo[];
}

export const getTodos = async (
  params?: GetTodosParams,
): Promise<GetTodosResponse> => {
  const response = await api.get("/todo", {
    params,
  });

  return response.data;
};

export const getTodoByID = async (todoID: string): Promise<Todo> => {
  const response = await api.get(`/todo/${todoID}`);

  return response.data;
};

export const createTodo = async (payload: CreateTodoPayload) => {
  const response = await api.post("/todo", payload);

  return response.data;
};

export const updateTodo = async (
  todoID: string,
  payload: UpdateTodoPayload,
) => {
  const response = await api.put(`/todo/${todoID}`, payload);

  return response.data;
};

export const deleteTodo = async (todoID: string) => {
  const response = await api.delete(`/todo/${todoID}`);

  return response.data;
};

export const markTodoCompleted = async (todoID: string) => {
  return updateTodo(todoID, {
    completed_at: new Date().toISOString(),
  });
};

export const markTodoPending = async (todoID: string) => {
  return updateTodo(todoID, {
    completed_at: null,
  });
};

File: backend/models/todo.go
package models

import "time"

// status
const (
	Completed  = "completed"
	Pending    = "pending"
	Incomplete = "incomplete"
	AdminRole  = "admin"
	//UserRole   = "user"
)

type Todo struct {
	ID          string     `db:"id" json:"id"`
	UserID      string     `db:"user_id" json:"user_id"`
	Name        string     `db:"name" json:"name"`
	Description string     `db:"description" json:"description"`
	PendingAt   *time.Time `db:"pending_at" json:"pending_at"`
	CompletedAt *time.Time `db:"completed_at" json:"completed_at"`
	CreatedAt   time.Time  `db:"created_at" json:"created_at"`
	//ArchivedAt  *time.Time `db:"archived_at" json:"archived_at"`
}

type CreateTodo struct {
	Name        string     `db:"name" json:"name" binding:"required"`
	Description string     `db:"description" json:"description" binding:"required"`
	PendingAt   *time.Time `db:"pending_at" json:"pending_at" binding:"required"`
}

type UpdateTodo struct {
	Name        *string    `db:"name" json:"name"`
	Description *string    `db:"description" json:"description"`
	PendingAt   *time.Time `db:"pending_at" json:"pending_at"`
	CompletedAt *time.Time `db:"completed_at" json:"completed_at"`
}

type CreateTodoForAllReq struct {
	Name        string     `json:"name" binding:"required"`
	Description string     `json:"description"`
	PendingAt   *time.Time `json:"pending_at"`
}
